src/common/libsdprocess/sdprocess.h

```diff
@@ -0,0 +1,158 @@
+
+#include <flux/core.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* sdprocess states, on changes, will lead to calls to callbacks in
+ * flux_sdprocess_state ().
+ *
+ * Possible state changes:
+ *
+ * init -> active
+ * init -> exited
+ * active -> exited
+ *
+ * Note that the active state can be missed, for example:
+ * - illegal command passed in by user, systemd exits and never runs
+ *   user command
+ * - systemd runs user process, but the command errors out before
+ *   flux_sdprocess_state_callback() is setup
+ * - libsdprocess cannot discern between the above, so the only state
+ *   that will be seen in the above two cases is "exited", the
+ *   "active" will not be seen.
+ */
+typedef enum {
+    FLUX_SDPROCESS_INIT = 0,    /* initial state */
+    FLUX_SDPROCESS_ACTIVE = 1,  /* systemd has run process */
+    FLUX_SDPROCESS_EXITED = 2,  /* systemd / process has exited */
+} flux_sdprocess_state_t;
+
+
+typedef struct flux_sdprocess flux_sdprocess_t;
+
+typedef void (*flux_sdprocess_state_f) (flux_sdprocess_t *sdp,
+                                        flux_sdprocess_state_t state,
+                                        void *arg);
+
+/* return > 0 to end iteration
+ * return 0 to continue
+ * return < 0 on error
+ */
+typedef int (*flux_sdprocess_list_f) (flux_t *h,
+                                      const char *unitname,
+                                      void *arg);
+
+/* Launch process under systemd
+ *
+ * command must be an absolute path
+ * argv & envv arrays are NULL terminated
+ * if unnecessary, set fd to < 0
+ *
+ * Setup of XDG_RUNTIME_DIR and DBUS_SESSION_BUS_ADDRESS environment
+ * variables are assumed.  If not, systemd will return error.
+ */
+flux_sdprocess_t *flux_sdprocess_exec (flux_t *h,
+                                       const char *unitname,
+                                       char **argv,
+                                       char **envv,
+                                       int stdin_fd,
+                                       int stdout_fd,
+                                       int stderr_fd);
+
+flux_sdprocess_t *flux_sdprocess_local_exec (flux_reactor_t *r,
+                                             const char *unitname,
+                                             char **argv,
+                                             char **envv,
+                                             int stdin_fd,
+                                             int stdout_fd,
+                                             int stderr_fd);
+
+/* Find process already launched under systemd
+ *
+ * Setup of XDG_RUNTIME_DIR and DBUS_SESSION_BUS_ADDRESS environment
+ * variables are assumed.  If not, systemd will return error.
+ */
+flux_sdprocess_t *flux_sdprocess_find_unit (flux_t *h,
+                                            const char *unitname);
+
+flux_sdprocess_t *flux_sdprocess_local_find_unit (flux_reactor_t *r,
+                                                  const char *unitname);
```

Comment:
These two functions seem to differ only in that one takes a `flux_t` handle and the other a `flux_reactor_t`.  How is one "local" and the other not?  Could the one that takes a `flux_t` handle be dropped and callers could just get the reactor from the handle themselves?

Same with `flux_sdprocess_local_exec()` and `flux_sdprocess_exec()`.

Reply:
I could be wrong but this is modeled after libsubprocess, which has `flux_exec(3)` and `flux_local_exec(3)`, which takes a reactor and therefore only invokes processes "locally", i.e. as a direct child of the calling process.

I could be wrong, but for `libsdprocess` the requirement for a function that takes a reactor instead of a handle is probably only for testing purposes, i.e. so that unit tests can be written without connecting to a broker. Since it isn't expected that the "local" functions will ever be used, maybe there is a better way to create the unit tests (using libtestutil?) that would allow @chu11 to remove the redundant functions and reduce the amount of maintained but unused code.

Reply:
I figured - just thought maybe the "local" part ought to be taken out of the API since everything is actually local.  But if there's a need I'm not seeing, it is not worth creating a kerfuffle.

Also: it looks like `flux_t` is only used internally so `sd_and_log_errno()` can call `flux_log( LOG_DEBUG)` when an error occurs.  I wonder if maybe the logging should be dropped or made an optional callback?  Meh, maybe it's fine.

Reply:
Yeah, it would be a good if we excise the local functions, only to have less code to maintain. I was having a similar thought, but then again this API isn't exported publicly and is only for internal use by job-exec..

Reply:
ahhh yes, unlike `libsubprocess` the `local` may not make as much sense.  Perhaps I should stick into a `_private.h` or similar.

Reply:
Or maybe the unit tests can just use `loopback_create()` from `libtestutil` to get a local `flux_t` and you can drop the duplicate functions? (Didn't check if that is feasible)